library/cwm/src/lib/cwm/tabs.rb

```diff
@@ -1,180 +1,76 @@
-
-    def handle(event)
-      # pass it to content of tab at first, maybe something stop passing
-      res = Yast::CWM.handleWidgets(@current_tab.cwm_widgets, event)
-      return res if res
-
-      new_id = event["ID"]
-      tab = tab_for_id(new_id)
-
-      return nil unless tab
-
-      return nil if @current_tab.widget_id == new_id
-
-      unless validate
-        mark_tab(@current_tab)
-        return nil
+  class Tabs < Pager
+    # {Tabs} does not have instances:
+    # {Tabs.new} overrides {Class.new} and calls
+    # either {DumbTabPager.new} or {PushButtonTabPager.new}.
+    def self.new(*args)
+      if Yast::UI.HasSpecialWidget(:DumbTab)
+        DumbTabPager.new(*args)
+      else
+        PushButtonTabPager.new(*args)
       end
-
-      store_tab(@current_tab.widget_id)
-
-      switch_tab(new_id)
-
-      nil
-    end
-
-    # store content of current tab
-    def store
-      store_tab(@current_tab.widget_id)
-    end
-
-    # validates current tab
-    def validate
-      Yast::CWM.validateWidgets(@current_tab.cwm_definition["widgets"], "ID" => @current_tab.widget_id)
     end
+  end
 
-    def help
-      @current_tab ? @current_tab.help : ""
-    end
+  # A {Pager} for the GUI, using the DumbTab widget
+  class DumbTabPager < Pager
+  #
 
   protected
 
-    # gets visual order of tabs
-    # This default implementation returns same order as passed to constructor
-    def tab_order
-      @tabs.map(&:widget_id)
+    # visually mark currently active tab
+    # @param page [Page]
+    def mark_page(page)
+      Yast::UI.ChangeWidget(Id(widget_id), :CurrentItem, page.widget_id)
     end
 
-    # stores tab with given id
-    def store_tab(tab_id)
-      Yast::CWM.saveWidgets(tab_for_id(tab_id).cwm_definition["widgets"], "ID" => tab_id)
+    def contents
+      panes = page_order.map do |page_id|
+        page = page_for_id(page_id)
+        Item(Id(page.widget_id), page.label, page.widget_id == initial_page_id)
+      end
+      DumbTab(Id(widget_id), panes, replace_point)
     end
+  end
 
-    # switch to target tab
-    def switch_tab(tab_id)
-      tab = tab_for_id(tab_id)
-      return unless tab
-
-      mark_tab(tab)
-      Yast::UI.ReplaceWidget(Id(replace_point_id), tab.cwm_definition["custom_widget"])
-      Yast::CWM.initWidgets(tab.cwm_definition["widgets"])
-      @current_tab = tab
+  # A {Pager} for ncurses, using PushButtons to simulate the tabs
+  class PushButtonTabPager < Pager
+  #
 
-      Yast::CWM.ReplaceWidgetHelp(widget_id, help)
-    end
+  protected
 
     # visually mark currently active tab
-    def mark_tab(tab)
-      if Yast::UI.HasSpecialWidget(:DumbTab)
-        Yast::UI.ChangeWidget(Id(widget_id), :CurrentItem, tab.widget_id)
-      else
-        if @current_tab
-          Yast::UI.ChangeWidget(
-            Id(@current_tab.widget_id),
-            :Label,
-            @current_tab.label
-          )
-        end
+    # @param page [Page]
+    def mark_page(page)
+      if @current_page
```

Comment:
It does, but the diff makes it really hard to see. Look here: https://github.com/yast/yast-yast2/blob/20d59b6d77c7e3e44f937ab1a2a45d67ae233715/library/cwm/src/lib/cwm/tabs.rb#L62